### PR TITLE
fix(observe): scope the latest observation and screenshot resources to one device (#6600)

### LINKED ISSUE #6600: observe(android|ios): observation/latest and observation/latest/screenshot resources can pair a hierarchy and screenshot from different devices
Child of [#6374](https://github.com/kaeawc/auto-mobile/issues/6374).

### Context

`getLatestObservation()` and `getLatestScreenshot()` in `src/server/observationResources.ts` back the `automobile:observation/latest` and `automobile:observation/latest/screenshot` MCP resources. Neither is scoped to a device — both resolve the "most recent" entry across every device the daemon is tracking, but from two independent stores with independent timestamps:

```ts
// observationResources.ts:146
const cachedResult = RealObserveScreen.getRecentCachedResult();
// -> ObserveScreen.ts:444-445 -> getObserveCacheStore().getRecentInMemory()
//    (FileSystemObserveCacheStore: scans cache across ALL devices for newest timestamp)

// observationResources.ts:126 / :188
const screenshotPath = RealObserveScreen.getRecentCachedScreenshotPath();
// -> ObserveScreen.ts:452-453 -> getScreenshotStateStore().getPath()
//    (ScreenshotStateRegistry: scans a SEPARATE Map<deviceId, ScreenshotState>
//     for the most-recently-updated device)
```

`getLatestScreenshot()` additionally falls back to `ScreenshotJobTracker.getMostRecentPendingDeviceId()` (observationResources.ts:207) when no cached screenshot path is found yet — again the globally most-recently-started pending job, not the job for the device whose hierarchy was just served.

Session-scoped resources in the same file already take the correct, device-scoped approach: `getRecentCachedResultForDevice(deviceId)` (observationResources.ts:443, 499) and `getRecentCachedScreenshotPathForDevice(deviceId)` (observationResources.ts:514).

### What goes wrong

1. Two devices are both under active observation by one daemon (e.g. an Android emulator and an iOS simulator, or two Android emulators in a multi-device session).
2. An `observe` call against device A caches a hierarchy at time T1 (`ObserveResultCacheStore.put`) and a screenshot path at time T1' via `ScreenshotStateStore.update`.
3. Shortly after, an `observe` call against device B caches only its hierarchy (T2 > T1), but its screenshot write lags (still in flight, or the screenshot job hasn't completed yet) so the screenshot store's most-recent entry is still device A's.
4. A client reads `automobile:observation/latest` (gets device B's hierarchy) and then `automobile:observation/latest/screenshot` (gets device A's screenshot, since device A is still the most-recently-updated entry in the screenshot store) — the two resources describe different devices, with no error or indication of the mismatch.

### Impact

In a multi-device automation session, an agent reading the "latest" resources to decide what UI is on screen can act on a screenshot that visually depicts a different device's state than the hierarchy it just parsed (or vice versa), producing incorrect taps/assertions with no diagnostic signal that a cross-device pairing occurred.

### How to reproduce or observe

A deterministic unit test can drive this without real devices, following the pattern in `test/server/observationResources.test.ts` (or the nearest existing suite for this file): construct two devices, populate the observe cache store for device B after populating the screenshot state store for device A only, then call `getLatestObservation()` and `getLatestScreenshot()` and assert their `deviceId`s match — the assertion should fail today because each resolves independently against its own store's "most recent" entry.

### Suggested direction

Have `getLatestObservation()` determine the device id from the resolved observation, then have `getLatestScreenshot()` (and its `getLatestScreenshotPath()` helper and the `ScreenshotJobTracker.getMostRecentPendingDeviceId()` fallback) accept/require that same device id and use the existing per-device accessors (`getRecentCachedResultForDevice`, `getRecentCachedScreenshotPathForDevice`) instead of the global "most recent across all devices" lookups. This mirrors the pattern already used for the session-scoped resources in the same file and avoids introducing a new locking/timeout primitive.

### How to verify the fix

A unit test asserting that `getLatestObservation()` and `getLatestScreenshot()` always resolve to the same `deviceId` when two devices have interleaved cache writes, added alongside the existing tests in `test/server/observationResources.test.ts`.

### Scope

- `src/server/observationResources.ts`: `getLatestObservation`, `getLatestScreenshot`, `getLatestScreenshotPath`
- `src/features/observe/ObserveScreen.ts`: `getRecentCachedResult`, `getRecentCachedScreenshotPath` (and their `ForDevice` counterparts)
- `src/features/observe/screenshot/ScreenshotStateRegistry.ts`: `InMemoryScreenshotStateStore.getPath`/`findLatest`
- `src/utils/ScreenshotJobTracker.ts`: `getMostRecentPendingDeviceId`

Related: none found.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by kaeawc**
## Problem

`automobile:observation/latest` and `automobile:observation/latest/screenshot` each ran their own unscoped "most recent across all devices" resolution against two stores that advance independently, plus a third global lookup for the pending-job fallback. When two devices are observed under one daemon and the second device's screenshot write lags its hierarchy write, a client reads device B's hierarchy and then device A's screenshot with no error, and can act on a screenshot depicting a different device.

## Fix

Resolve the device exactly once. `ObserveResultCacheStore.getRecentInMemoryEntry()` returns the newest cached result together with its owning device id, exposed as `RealObserveScreen.getRecentCachedObservation()`. The screenshot resource takes `deviceId` from that pair and uses the existing per-device accessors and a device-scoped `waitForCompletion`, mirroring the session-scoped resources in the same file. The unscoped accessors are deleted so the footgun cannot be reintroduced. Single-device behaviour is unchanged.

## Tests

Three deterministic unit tests with two devices: cross-device pairing is refused, the pending-job wait is scoped to the observed device, and the single-device path still serves its cached screenshot. The first two fail on main.

## Validation

typecheck gate clean, format:check clean, oxlint diff vs main empty, targeted 165 pass, broad server/db/daemon/utils 9506 pass / 0 fail, `test/features/observe` 2523 pass.

Closes [#6600](https://github.com/kaeawc/auto-mobile/issues/6600)

### CODE REVIEW

File: test/server/observationResources.test.ts [hunk 1]
```diff
@@ -535,3 +538,95 @@ describe("session screenshot resources", () => {
...
+    expect(screenshot.blob).toBeUndefined();
+    expect(screenshot.mimeType).toBe("application/json");
+    expect(JSON.parse(screenshot.text!).error).toContain("No screenshot available");
+  });
+
+  test("waits for the pending screenshot job of the observed device, not the newest job", async () => {
+    await cacheObservationFor(deviceA, "device-a-hierarchy");
+    getScreenshotStateStore().update(deviceA.deviceId, "/tmp/device-a.png");
+    await cacheObservationFor(deviceB, "device-b-hierarchy");
+    ScreenshotJobTracker.startJob(deviceB.deviceId, async () => {
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/server/observationResources.test.ts]: Correct — with only device B pending, the assertion was carried by the initial scoped lookup, not by the wait target. In 3447bc247 the test now registers two independently gated jobs, with device A's started last (so a global newest-pending lookup picks A) and never released, device A also holding a stale cached path, and the tracker on a `FakeTimer` so the 3s budget can't mask the wait. Only device B's gate opens; the read is raced against a microtask-drain sentinel. Verified by re-introducing `getMostRecentPendingDeviceId()`: the test now fails immediately with `still-pending`, where before it passed.

File: test/fakes/FakeObserveCacheStore.ts [hunk 2]
```diff
@@ -65,8 +68,9 @@ export class FakeObserveCacheStore implements ObserveResultCacheStore {
     return this.findMostRecent(deviceId);
   }
 
-  getRecentInMemory(): ObserveResult | undefined {
-    return this.findMostRecent();
+  getRecentInMemoryEntry(): RecentObserveCacheEntry | undefined {
+    const entry = this.findMostRecentEntry();
+    return entry ? { deviceId: entry.deviceId, result: entry.observeResult } : undefined;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/fakes/FakeObserveCacheStore.ts:73]: Implemented in a8ac44798. `FakeObserveCacheStore.findMostRecentEntry()` now uses `>=`, matching `FileSystemObserveCacheStore.collectLiveMostRecent()`, with a comment explaining that under a FakeTimer a shared tick is the common case rather than an edge case. Added `observeCacheRecentEntryParity.test.ts`, which drives the same A-then-B scenario through both stores; it reproduced the divergence exactly as reported before the fix and also pins the strictly-newer case.

File: test/server/observationResources.test.ts [hunk 3]
```diff
@@ -535,3 +556,122 @@ describe("session screenshot resources", () => {
...
+    registerObservationResources();
+    return ResourceRegistry.getResource(RESOURCE_URIS.LATEST_OBSERVATION)!.handler();
+  }
+
+  async function cacheObservationFor(device: BootedDevice, marker: string): Promise<void> {
+    const observeScreen = new RealObserveScreen(
+      device,
+      new FakeAdbClientFactory(new FakeAdbExecutor()),
+    );
+    await observeScreen.cacheObserveResult({
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/server/observationResources.test.ts:600]: Done in 96a2d43dc. `cacheObservationFor` now constructs `RealObserveScreen` with `{ cacheStore }` and a `FakeTimer`, which routes through `setObserveCacheStore`, so the resource handlers read an in-memory `FakeObserveCacheStore` and nothing in this block touches the shared cache directory. One store is created per test so both devices share it, and teardown calls `resetObserveCacheStore()` rather than `RealObserveScreen.clearCache()`. Re-verified against the pre-fix code path: both concurrency tests still fail there for their original reasons.

File: src/server/observationResources.ts [hunk 4]
```diff
@@ -185,8 +187,8 @@ async function getLatestObservation(): Promise<ResourceContent> {
 // Handler for latest screenshot resource (image/png as blob)
 async function getLatestScreenshot(): Promise<ResourceContent> {
   try {
-    const cachedResult = RealObserveScreen.getRecentCachedResult();
-    if (!cachedResult) {
+    const cachedObservation = RealObserveScreen.getRecentCachedObservation();
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/observationResources.ts]: Fixed properly in e6ad4e632. The hierarchy read now records the device whose observation it served, keyed by the reading client's device session, and the screenshot read serves that device instead of resolving the global latest a second time — falling back to the global latest only when the client hasn't read a hierarchy yet, or when the bound device's observation has since been evicted. The binding is per device rather than per snapshot because that's the finest identity available: the screenshot store keys its most recent capture by device id alone, with no snapshot id to bind to. New test `serves the screenshot of the observation the hierarchy read already returned` interleaves device A's observation between device B's two reads and asserts B's screenshot is still served; it fails against the previous code.

File: src/server/observationResources.ts [hunk 5]
```diff
@@ -120,10 +120,79 @@ export const RESOURCE_URIS = {
...
+
+// The device whose screenshot pairs with the hierarchy this client was last
+// served. Falls back to the globally latest observation when this client has
+// not read a hierarchy yet, or when the bound device's observation has since
+// been evicted or invalidated (its screenshot no longer describes anything this
+// client was shown).
+function resolveScreenshotDeviceId(context: ResourceReadContext | undefined): string | undefined {
+  const boundDeviceId = latestObservationBindings.get(bindingKey(context));
+  if (boundDeviceId && RealObserveScreen.getRecentCachedResultForDevice(boundDeviceId)) {
+    return boundDeviceId;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/observationResources.ts]: Fixed in 1ae0e6845 by removing the binding entirely rather than adding a consume-once TTL. `getLatestScreenshot` now resolves the device from a fresh `getRecentCachedObservation()` call on every read, so a screenshot-only reread after device B becomes latest always reflects the true current state — there is nothing left to go stale. The payload also gains `pairedScreenshotUri` (the session-scoped screenshot resource, or explicit `null`) for callers that need a guaranteed pair. New test: `re-resolves screenshot-only rereads after another device becomes latest`.

File: src/server/observationResources.ts [hunk 6]
```diff
@@ -120,10 +120,79 @@ export const RESOURCE_URIS = {
...
+// today's global-latest behavior and never worse than it.
+const UNSCOPED_BINDING_KEY = "__unscoped__";
+// Bindings are tiny (a device id per client) but the key space is unbounded
+// over a long-lived daemon's session churn, so keep only the most recent few.
+const MAX_LATEST_OBSERVATION_BINDINGS = 64;
+const latestObservationBindings = new Map<string, string>();
+
+function bindingKey(context: ResourceReadContext | undefined): string {
+  return context?.sessionUuid ?? UNSCOPED_BINDING_KEY;
+}
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/observationResources.ts]: Confirmed `ResourceReadContext` carries no per-connection identity for sessionless clients, so any binding-keyed fix would have had to invent one production cannot supply. Fixed in 1ae0e6845 by removing the binding map instead — there is no shared key for sessionless clients to collide on. Sessionless callers get point-in-time reads per resource and can follow `pairedScreenshotUri` when the served observation belongs to a device session. New test: `resolves sessionless interleaved reads from the current latest entry`, using literally `undefined` context for both callers.

File: src/server/observationResources.ts [hunk 7]
```diff
@@ -183,10 +257,10 @@ async function getLatestObservation(): Promise<ResourceContent> {
 }
 
 // Handler for latest screenshot resource (image/png as blob)
-async function getLatestScreenshot(): Promise<ResourceContent> {
+async function getLatestScreenshot(context?: ResourceReadContext): Promise<ResourceContent> {
   try {
-    const cachedResult = RealObserveScreen.getRecentCachedResult();
-    if (!cachedResult) {
+    const deviceId = resolveScreenshotDeviceId(context);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/observationResources.ts]: Fixed in 1ae0e6845 by making the screenshot read's device resolution independent of the hierarchy read's completion — each is its own atomic call into the same cache accessor, so there is no binding-not-yet-committed state to snapshot against. For a guaranteed match across two reads the hierarchy payload now advertises `pairedScreenshotUri` (session-scoped, stable per device by construction, verified to stay correct after another device observes in between). New test: `resolves a concurrent screenshot independently after a later observation lands`.

**Comment by kaeawc** [comment]
Contract after 1ae0e6845: the unscoped `automobile:observation/latest` and `automobile:observation/latest/screenshot` resources are independent, atomic, point-in-time reads — each resolves whatever device is globally most recent at the instant it is read, so a newer observation from another device landing between two reads can still leave the pair mismatched by design (there is no per-connection identity to bind on for sessionless clients). For a guaranteed match, the `latest` payload now carries a top-level `pairedScreenshotUri`: the exact `automobile:observation/session/{sessionUuid}/latest/screenshot` URI for the session that owns the served device, or explicit `null` when no session owns it. Clients that need the pair to match should follow that URI instead of reading the unscoped screenshot resource.

File: src/server/observationResources.ts [hunk 8]
```diff
@@ -160,11 +190,17 @@ async function getLatestObservation(): Promise<ResourceContent> {
       };
     }
 
+    const sessionUuid = resolvePairedScreenshotSessionUuid(cachedObservation.deviceId);
+    // Use null rather than omission when the latest observation has no active session owner.
+    const pairedScreenshotUri = sessionUuid
+      ? RESOURCE_URIS.SESSION_SCREENSHOT.replace("{sessionUuid}", sessionUuid)
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/observationResources.ts]: Agreed on all three counts, and the conclusion is that no existing resource can serve as the paired capture: the session-scoped screenshot is itself mutable latest-for-device, it is only readable by the owning session, and it can be read before the capture job finishes. `pairedScreenshotUri` is removed in c52aa7a79; the unscoped `latest` pair stays point-in-time per read (atomic per resource, no binding), and a guaranteed pairing is tracked as an observation-id-scoped screenshot resource in [#6940](https://github.com/kaeawc/auto-mobile/issues/6940), which quotes this thread.

[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/observationResources.ts]: Agreed on all three counts, and the conclusion is that no existing resource can serve as the paired capture: the session-scoped screenshot is itself mutable latest-for-device, it is only readable by the owning session, and it can be read before the capture job finishes. `pairedScreenshotUri` is removed in c52aa7a79; the unscoped `latest` pair stays point-in-time per read (atomic per resource, no binding), and a guaranteed pairing is tracked as an observation-id-scoped screenshot resource in [#6940](https://github.com/kaeawc/auto-mobile/issues/6940), which quotes this thread.

[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/observationResources.ts]: Agreed on all three counts, and the conclusion is that no existing resource can serve as the paired capture: the session-scoped screenshot is itself mutable latest-for-device, it is only readable by the owning session, and it can be read before the capture job finishes. `pairedScreenshotUri` is removed in c52aa7a79; the unscoped `latest` pair stays point-in-time per read (atomic per resource, no binding), and a guaranteed pairing is tracked as an observation-id-scoped screenshot resource in [#6940](https://github.com/kaeawc/auto-mobile/issues/6940), which quotes this thread.

**Comment by kaeawc** [comment]
Final contract after c52aa7a79: `automobile:observation/latest` and `automobile:observation/latest/screenshot` are independent, atomic, point-in-time reads with no cross-read binding (per-session bindings went stale, collided for sessionless clients, and raced on commit order; a session-scoped `pairedScreenshotUri` was mutable, not readable cross-session, and could precede the capture). The cross-device pairing bug from #6600 (B's hierarchy with A's screenshot from one client's sequential reads) is closed by the device-scoped screenshot resolution and the three scenario tests; a guaranteed same-capture pairing needs an observation-id-scoped resource, tracked in [#6940](https://github.com/kaeawc/auto-mobile/issues/6940).
